Source/WebKit/Platform/IPC/Connection.cpp

```diff
@@ -127,18 +152,45 @@ class Connection::SyncMessageState {
     };
     Deque<ConnectionAndIncomingMessage> m_messagesBeingDispatched; // Only used on the main thread.
     Deque<ConnectionAndIncomingMessage> m_messagesToDispatchWhileWaitingForSyncReply WTF_GUARDED_BY_LOCK(m_lock);
+
+    mutable std::atomic<unsigned> m_refCount { 1 };
 };
 
-Connection::SyncMessageState& Connection::SyncMessageState::singleton()
+Lock Connection::SyncMessageState::s_syncStateLock;
+
+DataMutexLocker<RefPtr<Connection::SyncMessageState>> Connection::syncState()
```

Comment:
So this DataMutex is just a convolution.
So you need to decide:
- if the reference to SyncMessageState is released in ~Connection(), then there's nothing to be locked.
- if the reference to SyncMessageState is released in Connection::invalidate(), then the incomingMessagesLock should protect the m_syncState.

Reply:
I think we agreed that it has to be released in `invalidate()`, since we need the `SerialFunctionDispatcher` pointer to go away entirely during `invalidate()` since it's not refcounted and the consumer of Connection needs to be able to immediately remove all references to it.

I'm not super convinced that sharing mutexes for more data members is simpler (especially since DataMutex mostly hides the mutex and just enforces compile-time locked access), but happy to make that change if you prefer.